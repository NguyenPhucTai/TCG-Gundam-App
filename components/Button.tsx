import React from 'react';
import { TouchableOpacity, Text, StyleSheet, ViewStyle, TextStyle } from 'react-native';

interface ButtonProps {
  title: string;
  onPress: () => void;
  disabled?: boolean;
  variant?: 'primary' | 'secondary';
  testID?: string;
}

export const Button: React.FC<ButtonProps> = ({
  title,
  onPress,
  disabled = false,
  variant = 'primary',
  testID,
}) => {
  const buttonStyle: ViewStyle = {
    ...styles.button,
    ...(variant === 'primary' ? styles.primaryButton : styles.secondaryButton),
    ...(disabled && styles.disabledButton),
  };

  const textStyle: TextStyle = {
    ...styles.buttonText,
    ...(variant === 'primary' ? styles.primaryText : styles.secondaryText),
    ...(disabled && styles.disabledText),
  };

  return (
    <TouchableOpacity
      style={buttonStyle}
      onPress={onPress}
      disabled={disabled}
      testID={testID}
      accessibilityRole="button"
      accessibilityState={{ disabled }}
    >
      <Text style={textStyle}>{title}</Text>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    paddingHorizontal: 20,
    paddingVertical: 12,
    borderRadius: 8,
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: 44,
  },
  primaryButton: {
    backgroundColor: '#007AFF',
  },
  secondaryButton: {
    backgroundColor: 'transparent',
    borderWidth: 1,
    borderColor: '#007AFF',
  },
  disabledButton: {
    backgroundColor: '#E5E5EA',
    borderColor: '#E5E5EA',
  },
  buttonText: {
    fontSize: 16,
    fontWeight: '600',
  },
  primaryText: {
    color: 'white',
  },
  secondaryText: {
    color: '#007AFF',
  },
  disabledText: {
    color: '#8E8E93',
  },
});
