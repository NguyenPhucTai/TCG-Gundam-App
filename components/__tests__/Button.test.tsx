import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react-native';
import { Button } from '../Button';

describe('Button Component', () => {
  const mockOnPress = jest.fn();

  beforeEach(() => {
    mockOnPress.mockClear();
  });

  it('renders correctly with required props', () => {
    render(<Button title="Test Button" onPress={mockOnPress} />);
    
    expect(screen.getByText('Test Button')).toBeTruthy();
  });

  it('calls onPress when pressed', () => {
    render(<Button title="Test Button" onPress={mockOnPress} testID="test-button" />);
    
    const button = screen.getByTestId('test-button');
    fireEvent.press(button);
    
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });

  it('does not call onPress when disabled', () => {
    render(
      <Button
        title="Disabled Button"
        onPress={mockOnPress}
        disabled={true}
        testID="disabled-button"
      />
    );
    
    const button = screen.getByTestId('disabled-button');
    fireEvent.press(button);
    
    expect(mockOnPress).not.toHaveBeenCalled();
  });

  it('renders primary variant by default', () => {
    render(<Button title="Primary Button" onPress={mockOnPress} testID="primary-button" />);
    
    const button = screen.getByTestId('primary-button');
    expect(button).toBeTruthy();
  });

  it('renders secondary variant when specified', () => {
    render(
      <Button
        title="Secondary Button"
        onPress={mockOnPress}
        variant="secondary"
        testID="secondary-button"
      />
    );
    
    const button = screen.getByTestId('secondary-button');
    expect(button).toBeTruthy();
  });

  it('has correct accessibility properties', () => {
    render(<Button title="Accessible Button" onPress={mockOnPress} testID="accessible-button" />);
    
    const button = screen.getByTestId('accessible-button');
    expect(button.props.accessibilityRole).toBe('button');
    expect(button.props.accessibilityState.disabled).toBe(false);
  });

  it('has correct accessibility state when disabled', () => {
    render(
      <Button
        title="Disabled Button"
        onPress={mockOnPress}
        disabled={true}
        testID="disabled-accessible-button"
      />
    );
    
    const button = screen.getByTestId('disabled-accessible-button');
    expect(button.props.accessibilityState.disabled).toBe(true);
  });

  it('matches snapshot for primary variant', () => {
    const tree = render(<Button title="Primary Button" onPress={mockOnPress} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('matches snapshot for secondary variant', () => {
    const tree = render(
      <Button title="Secondary Button" onPress={mockOnPress} variant="secondary" />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('matches snapshot for disabled state', () => {
    const tree = render(<Button title="Disabled Button" onPress={mockOnPress} disabled={true} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
