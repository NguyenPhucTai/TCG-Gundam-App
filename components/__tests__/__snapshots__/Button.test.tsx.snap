// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component matches snapshot for disabled state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#E5E5EA",
      "borderColor": "#E5E5EA",
      "borderRadius": 8,
      "justifyContent": "center",
      "minHeight": 44,
      "opacity": 1,
      "paddingHorizontal": 20,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      {
        "color": "#8E8E93",
        "fontSize": 16,
        "fontWeight": "600",
      }
    }
  >
    Disabled Button
  </Text>
</View>
`;

exports[`Button Component matches snapshot for primary variant 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#007AFF",
      "borderRadius": 8,
      "justifyContent": "center",
      "minHeight": 44,
      "opacity": 1,
      "paddingHorizontal": 20,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 16,
        "fontWeight": "600",
      }
    }
  >
    Primary Button
  </Text>
</View>
`;

exports[`Button Component matches snapshot for secondary variant 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "#007AFF",
      "borderRadius": 8,
      "borderWidth": 1,
      "justifyContent": "center",
      "minHeight": 44,
      "opacity": 1,
      "paddingHorizontal": 20,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      {
        "color": "#007AFF",
        "fontSize": 16,
        "fontWeight": "600",
      }
    }
  >
    Secondary Button
  </Text>
</View>
`;
